distributed/worker.py

```diff
@@ -3226,14 +3238,28 @@ def apply_function_actor(
     thread_state.execution_state = execution_state
     thread_state.key = key
 
+    notify_task_started(plugins, execution_state['worker'], args, kwargs)
     result = function(*args, **kwargs)
+    notify_task_finished(plugins, execution_state['worker'], result)
 
     with active_threads_lock:
         del active_threads[ident]
 
     return result
 
 
+def notify_task_started(plugins, worker, args, kwargs):
+    for plugin in plugins.values():
+        if hasattr(plugin, "task_started"):
+            plugin.task_started(worker, args, kwargs)
+
+
+def notify_task_finished(plugins, worker, result):
+    for plugin in plugins.values():
+        if hasattr(plugin, "task_finished"):
+            plugin.task_finished(worker, result)
```

Comment:
might be helpful to have the result state (successfully vs exception) reported to the plugin as well

Reply:
That's a good idea, but two points:

1. it would basically mean adding try/catches at every call site, which would be quite invasive.

2. maybe it could be also be done by adding a general exception handling hook to the worker plugin.

This also makes me think that we also might want to add the task key to each of these function calls. Otherwise it would make it hard to connect the dots between different logs.

Reply:
> That's a good idea, but two points ...

Then let's clarify the behavior: if a task crashes, what do we expect to happen? Does the worker plugin get:

1. a `task_done` notification and an (TBD) exception/crashed notification
2. only an (TBD) exception/crashed notification
3. only a `done` notification

> This also makes me think that we also might want to add the task key to each of these function calls.

IMHO that would be extremely helpful.

Reply:
> Then let's clarify the behavior: if a task crashes, what do we expect to happen?

That was my question as well ;) It's really just a question of API design, and even for the user it really doesn't make much of a difference I would say. I'll implement 3. and then we can discuss further.

> IMHO that would be extremely helpful.

This is now part of the PR.